packages/@lwc/ssr-runtime/src/clone-and-deep-freeze.ts

```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2024, Salesforce, Inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+
+import { freeze, entries, isObject, isArray, create, isNull } from '@lwc/shared';
+
+// Deep freeze and clone an object. Designed for cloning/freezing child props when passed from a parent to a child so
+// that they are immutable. This is one of the normal guarantees of both engine-dom and engine-server that we want to
+// emulate in ssr-runtime. The goal here is that a child cannot mutate the props of its parent and thus affect
+// the parent's rendering, which would lead to bidirectional reactivity and mischief.
+export function cloneAndDeepFreeze(obj: any): any {
+    if (isArray(obj)) {
+        const res = [];
+        for (let i = 0; i < obj.length; i++) {
+            res[i] = cloneAndDeepFreeze(obj[i]);
+        }
+        freeze(res);
+        return res;
+    } else if (!isNull(obj) && isObject(obj)) {
+        const res = create(null);
+        for (const [key, value] of entries(obj as any)) {
```

Comment:
```suggestion
    } else if (isObject(obj) && !isNull(obj)) {
        const res = create(null);
        for (const [key, value] of entries(obj)) {
```
`isNull` is a type guard for `null`, but `isObject` is actually a guard for `object | null` (see #4003). By checking `!isNull` and then `isObject`, we're kinda saying "it's not null, but actually it could be?". So we end up with `object | null` within the block. But `entries` just wants `object`, which is why it complains. If we switch to `isObject` then `!isNull`, we're checking "is object or null, and it's not null", and we get what we want, just `object`.

Reply:
As a bonus typescript feature, we also could avoid `as any` here by using a non-null assertion, `obj!`. That changes the type to just remove `null` and `undefined`, rather than obliterating everything.